From 17735a7c33aa59ff2e430dde9e9686110d6db3f4 Mon Sep 17 00:00:00 2001
From: Lorenzo Pirritano <lpirritano@anaconda.com>
Date: Tue, 23 Jul 2024 11:33:44 +0200
Subject: [PATCH 3/3] n_restarts_optimizer must be positive

`scikit-learn` introduced a guard for this parameter in version `1.4.0`,
before that, one could do it with `-1`. Now it must start from 0.

The patch introduces behavior change for `smac` in this part of the code.
Tests more or less pass. The 0 means make 1 pass from the documentation,
but we can't know possibly the effects. 

> Note that n_restarts_optimizer == 0 implies that one run is performed.

https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.GaussianProcessRegressor.html

In the source code, there is a check for that (version `1.5.1`):
https://github.com/scikit-learn/scikit-learn/blob/70fdc843a4b8182d97a3508c1a426acc5e87e980/sklearn/gaussian_process/_gpr.py#L193

We could pin the upper bound to use `scikit-learn` version `1.3.2`
because the check was introduced in `1.4.0`:
https://github.com/scikit-learn/scikit-learn/blob/093e0cf14aff026cca6097e8c42f83b735d26358/sklearn/gaussian_process/_gpr.py#L206

However, looking at the code, `1.3.2` and `1.5.1` don't use values `n_restarts_optimizer<=0`.

1.3.2:
https://github.com/scikit-learn/scikit-learn/blob/093e0cf14aff026cca6097e8c42f83b735d26358/sklearn/gaussian_process/_gpc.py#L236-L247

1.5.1:
https://github.com/scikit-learn/scikit-learn/blob/70fdc843a4b8182d97a3508c1a426acc5e87e980/sklearn/gaussian_process/_gpc.py#L236-L247

it does not look like `n_restarts_optimizer` is used anywhere else,
but of course, we can't guarantee this 100% or that `smac` or any 
caller/user/plugin of the library will use its value.

---
 smac/epm/gaussian_process.py      | 2 +-
 smac/epm/gaussian_process_mcmc.py | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/smac/epm/gaussian_process.py b/smac/epm/gaussian_process.py
index 81e1dcd..5efb5c1 100644
--- a/smac/epm/gaussian_process.py
+++ b/smac/epm/gaussian_process.py
@@ -144,7 +144,7 @@ class GaussianProcess(BaseModel):
             kernel=self.kernel,
             normalize_y=False,
             optimizer=None,
-            n_restarts_optimizer=-1,  # Do not use scikit-learn's optimization routine
+            n_restarts_optimizer=0,  # Do not use scikit-learn's optimization routine
             alpha=0,  # Governed by the kernel
             random_state=self.rng,
         )
diff --git a/smac/epm/gaussian_process_mcmc.py b/smac/epm/gaussian_process_mcmc.py
index c77f198..038e07b 100644
--- a/smac/epm/gaussian_process_mcmc.py
+++ b/smac/epm/gaussian_process_mcmc.py
@@ -292,7 +292,7 @@ class GaussianProcessMCMC(BaseModel):
             kernel=self.kernel,
             normalize_y=False,
             optimizer=None,
-            n_restarts_optimizer=-1,  # Do not use scikit-learn's optimization routine
+            n_restarts_optimizer=0,  # Do not use scikit-learn's optimization routine
             alpha=0,  # Governed by the kernel
         )
 
-- 
2.39.1

